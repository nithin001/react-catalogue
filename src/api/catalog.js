import instance from './base';
import Ajv from 'ajv';

export const get = () => new Promise((resolve, reject) => {
  instance.get('/catalog')
    .then((response) => {
      if (response.data === undefined) {
        reject('backend_error');
        return;
      }
      const data = response.data;
      const ajv = new Ajv();
      const validate = ajv.compile(schema);
      const valid = validate(data);
      if (valid) {
        resolve(data);
      } else {
        reject('backend_error');
      }
    })
    .catch((error) => {
      reject('backend_error');
    });
});

const schema = {
  $id: 'http://challenge.monoqi.net/catalog',
  type: 'object',
  definitions: {},
  $schema: 'http://json-schema.org/draft-07/schema#',
  additionalProperties: false,
  properties: {
    articles: {
      $id: '/properties/articles',
      type: 'array',
      items: {
        $id: '/properties/articles/items',
        type: 'object',
        additionalProperties: false,
        properties: {
          sku: {
            $id: '/properties/articles/items/properties/sku',
            type: 'string',
            title: 'The Sku Schema ',
            default: '',
          },
          name: {
            $id: '/properties/articles/items/properties/name',
            type: 'string',
            title: 'The Name Schema ',
            default: '',
          },
          image: {
            $id: '/properties/articles/items/properties/image',
            type: 'string',
            title: 'The Image Schema ',
            default: '',
          },
          price: {
            $id: '/properties/articles/items/properties/price',
            type: 'object',
            additionalProperties: false,
            properties: {
              amount: {
                $id: '/properties/articles/items/properties/price/properties/amount',
                type: 'number',
                title: 'The Amount Schema ',
                default: 0,
              },
              currency: {
                $id: '/properties/articles/items/properties/price/properties/currency',
                type: 'string',
                title: 'The Currency Schema ',
                default: '',
              },
            },
            required: [
              'amount',
              'currency',
            ],
          },
        },
        required: [
          'sku',
          'name',
          'image',
          'price',
        ],
      },
    },
  },
  required: [
    'articles',
  ],
};
