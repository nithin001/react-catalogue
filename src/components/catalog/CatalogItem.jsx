import React from 'react';

export const CatalogItem = ({ item, addToCart }) => (
  <div className={'catalog-item'}>
    <span className={'name'}>{item.get('name')}</span>
    <img src={item.get('image')} className={'image'}/>
    <span className={'price-amount'}>{item.getIn(['price', 'amount'])}</span>
    <span className={'price-currency'}>{item.getIn(['price', 'currency'])}</span>
    <button onClick={() => {addToCart(item.get('sku'));}}>Add to cart</button>
  </div>);

