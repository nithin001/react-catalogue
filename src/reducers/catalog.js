import Immutable from 'immutable';
const catalog = (state = Immutable.Map(), action) => {
  switch (action.type) {
    case 'POPULATE_CATALOG': {
      const articles = action.payload.articles;
      return Immutable.fromJS(articles);
    }
  }
  return state;
};
export default catalog;