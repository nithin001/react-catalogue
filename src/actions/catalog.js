import { get } from '../api/catalog';

export const load = () => (dispatch) => {
  dispatch({ type: 'LOAD_CATALOG_START' });
  return get()
    .then((catalog) => {
      dispatch({ type: 'POPULATE_CATALOG', payload: catalog });
      dispatch({ type: 'LOAD_CATALOG_COMPLETE' });
    })
    .catch(() => {
      dispatch({ type: 'LOAD_ERROR' });
    });
};
