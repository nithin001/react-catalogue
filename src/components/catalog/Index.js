import React from 'react';
import { connect } from 'react-redux';
import { CatalogItem } from './CatalogItem';
export const Catalog = ({ items, addToCart }) => {
  const articles = [];
  items.keySeq().forEach((item) => {
    articles.push(items.get(item));
  });
  const articlesList = articles.map(item => {
    return (<CatalogItem key={item.get('sku')} item={item} addToCart={addToCart}/>);
  });
  return (<div className={'catalog'}>
    {articlesList}
  </div>);
};

export const mapStateToProps = (state, ownProps) => {
  return { catalog: state.catalog };
};

export const mapDispatchToProp = (dispatch) => ({
    addToCart: (sku) => dispatch(addToCart(sku)),
  }
);

export default connect(mapStateToProps, mapDispatchToProp)(Catalog);


