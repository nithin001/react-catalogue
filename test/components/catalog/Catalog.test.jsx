import React from 'react';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { MemoryRouter } from 'react-router';
import { Catalog, mapStateToProps } from '../../../src/components/catalog';
import catalog from '../../../src/reducers/catalog';
configure({ adapter: new Adapter() });
var mockCatalog = {
  'articles': [{
    'sku': '199203',
    'name': 'IPad',
    'image': 'https://picsum.photos/480/480/?image=6',
    'price': { 'amount': 229, 'currency': 'EUR' }
  }]
};

const catalogState = catalog(undefined, { type: 'POPULATE_CATALOG', payload: mockCatalog });

describe('Catalog component', () => {
  it('should render the list of CatalogItems', () => {
    const addToCart = jest.fn();
    const props = { catalog: catalogState, addToCart };
    const catalogItems = renderer.create(<MemoryRouter><Catalog
      {...props}
    /></MemoryRouter>);
    expect(catalogItems).toMatchSnapshot();
  });
});

describe('Connected Catalog component', () => {
  it('should map state values to props of the component', () => {
    const state = {
      catalog: catalogState,
    };
    const result = mapStateToProps(state);
    expect(result).toEqual({ catalog: catalogState });
  });
});