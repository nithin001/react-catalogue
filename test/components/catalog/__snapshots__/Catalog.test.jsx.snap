// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog component should render the list of CatalogItems 1`] = `
<div
  className="catalog"
>
  <div
    className="catalog-item"
  >
    <span
      className="name"
    >
      IPad
    </span>
    <img
      className="image"
      src="https://picsum.photos/480/480/?image=6"
    />
    <span
      className="price-amount"
    >
      229
    </span>
    <span
      className="price-currency"
    >
      EUR
    </span>
    <button
      onClick={[Function]}
    >
      Add to cart
    </button>
  </div>
</div>
`;
